### PR TITLE
Add Links support

### PR DESCRIPTION
**Pull request by pakrym**
@lmolkova 

Fixes: https://github.com/Azure/azure-sdk-for-net/issues/7552

### CODE REVIEW

File: src/OpenTelemetry.Collector.Dependencies/AzureSdkDiagnosticListener.cs [hunk 1]
```diff
@@ -64,6 +67,13 @@ public override void OnStartActivity(Activity current, object valueValue)
 
             var span = spanBuilder.StartSpan();
 
+            var links = LinksPropertyFetcher.Fetch(valueValue) as IEnumerable<Activity> ?? Array.Empty<Activity>();
+
+            foreach (var link in links)
+            {
+                span.AddLink(Link.FromActivity(link));
```
Comment (lmolkova) [inline src/OpenTelemetry.Collector.Dependencies/AzureSdkDiagnosticListener.cs]: please add them before ` var span = spanBuilder.StartSpan()`

File: src/OpenTelemetry.Collector.Dependencies/AzureSdkDiagnosticListener.cs [hunk 2]
```diff
@@ -60,6 +63,13 @@ public override void OnStartActivity(Activity current, object valueValue)
...
                 .SetSampler(this.sampler);
 
+            var links = LinksPropertyFetcher.Fetch(valueValue) as IEnumerable<Activity> ?? Array.Empty<Activity>();
+
+            foreach (var link in links)
+            {
+                spanBuilder.AddLink(Link.FromActivity(link));
+            }
+
             spanBuilder.SetSpanKind(isHttp ? SpanKind.Client : SpanKind.Internal);
```
Comment (lmolkova) [inline src/OpenTelemetry.Collector.Dependencies/AzureSdkDiagnosticListener.cs]: this doesn't look right, it should be based on tag

File: src/OpenTelemetry.Collector.Dependencies/AzureSdkDiagnosticListener.cs [hunk 3]
```diff
@@ -60,6 +63,13 @@ public override void OnStartActivity(Activity current, object valueValue)
                 .SetCreateChild(false)
                 .SetSampler(this.sampler);
 
+            var links = LinksPropertyFetcher.Fetch(valueValue) as IEnumerable<Activity> ?? Array.Empty<Activity>();
+
+            foreach (var link in links)
+            {
+                spanBuilder.AddLink(Link.FromActivity(link));
```
Comment (lmolkova) [inline src/OpenTelemetry.Collector.Dependencies/AzureSdkDiagnosticListener.cs]: I believe Link .FromActivity does not work correctly - it assumes Activity is started (id format is set) and also uses Activity SpanId. In our case (which seems typical for links) Activity ParentSpanId is one that we care about and Activity is not started.

I'll try to fix it today.

File: src/OpenTelemetry.Collector.Dependencies/AzureSdkDiagnosticListener.cs (hunk 2 above)
Comment (pakrym) [reply src/OpenTelemetry.Collector.Dependencies/AzureSdkDiagnosticListener.cs]: done

File: src/OpenTelemetry.Collector.Dependencies/AzureSdkDiagnosticListener.cs (hunk 3 above)
Comment (pakrym) [reply src/OpenTelemetry.Collector.Dependencies/AzureSdkDiagnosticListener.cs]: What does starting an activity change in terms of Ids?

Comment (lmolkova) [reply src/OpenTelemetry.Collector.Dependencies/AzureSdkDiagnosticListener.cs]: it generates new span id, it also determines format id during start as far as i remember

Comment (pakrym) [reply src/OpenTelemetry.Collector.Dependencies/AzureSdkDiagnosticListener.cs]: we set id format in the producer to always be W3C but I see your point with span id.

**Comment by pakrym** [comment]
Is there a good way to see links exported. As far as I understand they are not supported in Zipkin. Does Azure Monitor display the links?

**Comment by lmolkova** [comment]
> Is there a good way to see links exported. As far as I understand they are not supported in Zipkin. Does Azure Monitor display the links?

AFAIK no UX in OSS backends.
Azure monitor now just sends them in custom properties - we are working on UX behind it and schema will be different.  So the only way I know is just to look into custom dimensions and compare ids.

File: src/OpenTelemetry.Collector.Dependencies/AzureSdkDiagnosticListener.cs (hunk 3 above)
Comment (lmolkova) [reply src/OpenTelemetry.Collector.Dependencies/AzureSdkDiagnosticListener.cs]: I've thought about it a bit more. Considering different ways `Link.FromActivity` should behave (use parent of new context), I want to remove this API at all. And call `Link.FromSpanContext` that does not have this ambiguity. If we'll ever need `FromActivity` we'll return it back in some form.

So here, would you mind explicitly calling `Link.FromSpanContext(SpanContext.Create(link.TraceId, link.ParentSpanId, link.ActivityTraceFlags, Tracestate.Empty))` ?

**Comment by lmolkova** [comment]
@pakrym is it ready for merge? (it's still draft)

**Comment by pakrym** [comment]
I still want to verify it E2E and that's not done yet.

**Comment by lmolkova** [comment]
@pakrym it's not a draft anymore, should I merge it?

**Comment by pakrym** [comment]
Yes

**Pull request merged**